setuptools-rust-hook: Use correct python for cross

kjeremy opened the pull request:
## Description of changes

Sets `PYO3_PYTHON` according to https://pyo3.rs/v0.20.0/building_and_distribution#cross-compiling

Setting `PYO3_INCLUDE_DIR` doesn't seem to be necessary for fixing #262073 but I set it for good measure.

Fixes #262073

@Artturin 
@Cynerd

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

kjeremy committed:
setuptools-rust-hook: Use correct python for cross

SuperSandro2000 reviewed (approved):
not to fit on cross things but looks logical to me

Artturin commented:
Has to target staging

kjeremy commented:
> Has to target staging

I changed the target to staging.

kjeremy commented:
The ofborg-eval failure doesn't look related to this PR.

Artturin commented:
@ofborg eval

kjeremy commented:
@ofborg eval

Artturin commented:
Gotta force-push

amjoseph-nixpkgs commented:
I cherry-picked this back to `master` but it's still trying to run `hostPlatform` binaries:

```
declare -x PYO3_CROSS_INCLUDE_DIR="/nix/store/pzlcmpd2fjc401cjsfx3b424mdl5w86b-python3-armv7l-unknown-linux-gnueabihf-3.11.5/include/python3.11"
declare -x PYO3_CROSS_LIB_DIR="/nix/store/pzlcmpd2fjc401cjsfx3b424mdl5w86b-python3-armv7l-unknown-linux-gnueabihf-3.11.5/lib/python3.11"
declare -x PYO3_PYTHON="/nix/store/pzlcmpd2fjc401cjsfx3b424mdl5w86b-python3-armv7l-unknown-linux-gnueabihf-3.11.5/bin/python3.11"
```

Is there something in `staging` but not yet in `master` that makes this PR work?

Also, note that `setuptoolsRustBuildHook.__spliced` doesn't exist.  Something is really borked here.

Artturin commented:
> Also, note that setuptoolsRustBuildHook.__spliced doesn't exist. Something is really borked here.

It's this https://github.com/NixOS/nixpkgs/pull/228139
The hooks aren't spliced withing the sets `callPackage` but they will be spliced outside of it

`pkgsCross.aarch64-multiplatform.__splicedPackages.python3Packages.setuptoolsRustBuildHook.__spliced` is spliced

```
nix-repl> pkgsCross.aarch64-multiplatform.__splicedPackages.python3.pkgs.setuptoolsRustBuildHook.__spliced
error: attribute '__spliced' missing
```
is the fault of https://github.com/NixOS/nixpkgs/issues/211340

Artturin commented:
@kjeremy do you have binfmt enabled?

Artturin commented:
Could work around it by doing `pyBin = "python${python.pythonVersion}";`
but then
`#error "LONG_BIT definition appears wrong for platform (bad gcc/glibc config?)."`
for some reason

amjoseph-nixpkgs commented:
> Could work around it by doing `pyBin = "python${python.pythonVersion}";`

Eeeew.

> but then
> `#error "LONG_BIT definition appears wrong for platform (bad gcc/glibc config?)."`
> for some reason, even though

Can we just move all the python hooks out of `hooks` and into `pythonPackages` so they get spliced?  Otherwise we're going to keep having problems like this :(

kjeremy commented:
> `declare -x PYO3_PYTHON="/nix/store/pzlcmpd2fjc401cjsfx3b424mdl5w86b-python3-armv7l-unknown-linux-gnueabihf-3.11.5/bin/python3.11"` should be the build python I think
> 
> ```
> nix-repl> p = lib.elemAt pkgsCross.aarch64-multiplatform.__splicedPackages.python3.pkgs.cryptography.nativeBuildInputs 10
> nix-repl> p.stdenv.hostPlatform.system
> "aarch64-linux"
> ```
> 
> @kjeremy do you have binfmt enabled?
> 
> `python.pythonForBuild.pkgs.setuptoolsRustBuildHook` would make hostPlatform correct, but then the vars aren't for the target
> 
> ```
> declare -x PYO3_CROSS_INCLUDE_DIR="/nix/store/ffll6glz3gwx342z0ch8wx30p5cnqz1z-python3-3.11.5/include/python3.11"
> declare -x PYO3_CROSS_LIB_DIR="/nix/store/ffll6glz3gwx342z0ch8wx30p5cnqz1z-python3-3.11.5/lib/python3.11"
> declare -x PYO3_PYTHON="/nix/store/ffll6glz3gwx342z0ch8wx30p5cnqz1z-python3-3.11.5/bin/python3.11"
> ```

Yes I have binfmt enabled.

Commit:
stdenv: lib.warn if drv.__splice is missing when expected

This commit adds a `lib.warn` check to stdenv.mkDerivation, causing
it to emit a warning if a dependency which ought to be spliced in
fact is not.

kjeremy commented:
Not sure what the right fix is here.

amjoseph-nixpkgs commented:
> Not sure what the right fix is here.

We probably need to deal with https://github.com/NixOS/nixpkgs/issues/263019 first.

Commit:
gcc: use stdenv.cc.bintools instead of targetPackages_bintools in buildInputs

Commit:
isl: fix splicing failure

Commit:
stdenv: allow unspliced hostTarget dependencies (for now)

Commit:
stdenv: add list-of-shame

Co-authored-by: Artturi <Artturin@artturin.com>

Commit:
rustc-wasm32: give it a custom pname for stdenv list-of-shame

Commit:
cracklib: do not use buildPackages for self-reference (fails to splice)

amjoseph-nixpkgs commented:
Rebase on

- https://github.com/NixOS/nixpkgs/pull/263082

and then run

```
nix-instantiate . -A pkgsCross.armv7l-hf-multiplatform.python3Packages.cryptography
```

you will get

```
trace: warning: derivation ncurses:
                     unspliced build-host dependency ncurses
                       build=x86_64-unknown-linux-gnu
                       target=armv7l-unknown-linux-gnueabihf

trace: warning: derivation python3:
                     unspliced build-host dependency python3
                       build=x86_64-unknown-linux-gnu
                       target=armv7l-unknown-linux-gnueabihf

trace: warning: derivation !!no pname!!:
                     unspliced build-host dependency cargo-vendor-normalise
                       build=x86_64-unknown-linux-gnu
                       target=armv7l-unknown-linux-gnueabihf

trace: warning: derivation !!no pname!!:
                     unspliced build-host dependency armv7l-unknown-linux-gnueabihf-cargo-1.72.1
                       build=x86_64-unknown-linux-gnu
                       target=armv7l-unknown-linux-gnueabihf

trace: warning: derivation python3:
                     unspliced build-host dependency python3
                       build=x86_64-unknown-linux-gnu
                       target=armv7l-unknown-linux-gnueabihf

trace: warning: derivation cryptography:
                     unspliced build-host dependency setuptools-rust-setup-hook
                       build=x86_64-unknown-linux-gnu
                       target=armv7l-unknown-linux-gnueabihf

trace: warning: derivation python3:
                     unspliced build-host dependency python3
                       build=x86_64-unknown-linux-gnu
                       target=armv7l-unknown-linux-gnueabihf
```

So, those are the bugs we need to fix.

Commit:
targetPackages_bintools: give it a top-level entry

In order to avoid setting off the unspliced-dependency check, we
need to make sure that the targetPackages.stdenv.cc.bintools used by
gcc gets spliced.  In order for it to be spliced, it needs a
top-level entry in all-packages.nix.  Therefore, this commit adds one.

This also causes us to stop abusing targetPackages to access the
linker, which is a good thing.

Commit:
stdenv: alphabetize list-of-shame

Commit:
stdenv: improve diagnostics for unspliced-dependency check

Commit:
Revert "gcc: use stdenv.cc.bintools instead of targetPackages_bintools in buildInputs"

This reverts commit 0899b6b8746790eec9dc49ea4f4a6c3def072e3b.

SuperSandro2000 commented:
@kjeremy can you do the fix noticed in the comment before this one?

kjeremy commented:
> @kjeremy can you do the fix noticed in the comment before this one?

Do you mean rebase on https://github.com/NixOS/nixpkgs/pull/263082? If so then yes I can do that. Note that my solution to this isn't correct because I had `binfmt` enabled.

kjeremy commented:
@SuperSandro2000 I've rebased and did a force push

SuperSandro2000 commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -283,32 +347,89 @@ else let
+           !(list-of-shame ? ${approximate-pname})
+          )
+
+            # Hi there!  If you're reading this, it's probably
```
This would be way easier to read and a lot less lines if we'd wrote one setence per line.

kjeremy replied:
Yes it would be but that's part of #263082. I feel like if that got merged it would force us to address the cross-compilation issues.

kjeremy commented:
cryptography seems to build for me now
